core/src/main/scala/kafka/server/DelayedOperation.scala

```diff
@@ -119,7 +119,12 @@ abstract class DelayedOperation(delayMs: Long,
   /**
    * Thread-safe variant of tryComplete()
    */
-  private[server] def safeTryComplete(): Boolean = inLock(lock)(tryComplete())
+  private[server] def safeTryComplete(): Boolean = inLock(lock) {
+    if (isCompleted)
+      true
```

Comment:
> If yes execute the completion logic by calling
>  forceComplete() and return true iff forceComplete returns true; otherwise return false

This is the return value definition. So if the request is completed, we should return `false`.

Reply:
Yeah, you are right. As the request is already completed then it should return false as other thread should have already bumped the `completed` count. And it should be removed from iteration once watched sees it completed. I made the change.